lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py

```diff
@@ -0,0 +1,296 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+# TODO
+# - make unit tests
+# - test what happens when the first host is down, goes to second
+# - test what happens if all hosts are down
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+
+DOCUMENTATION = '''
+---
+module: vmware_datastore_upload
+short_description: Upload a local file to a VMware datastore.
+description:
+    - Upload a local file to a VMware datastore with the option to
+      skip the upload if the file already exists and is the same.
+author:
+    - "James E. King III (@jeking3) <jking@apache.org>"
+version_added: 2.8
+notes:
+    - Tested on vSphere 6.7
+requirements:
+    - python >= 2.6
+    - PyVmomi
+    - requests >= 2.18.4
+options:
+    datacenter:
+      description:
+      - Name of the datacenter containing the datastore.
+      - If not specified, and more than one datastore with the
+        same name exists, the action will fail.
+      required: no
+    datastore:
+      description:
+      - Name of datastore that will store the file.
+      required: yes
+    source:
+      description:
+      - The full path to the local file to push to vCenter.
+      required: yes
+    destination:
+      description:
+      - The full path to the destination file in the datastore.
+      required: yes
+    debounce:
+      description:
+      - Uploading large files unnecessarily can be avoided
+        with the C(debounce) option.
+      - C(hash) will hash C(source) (using SHA256) and check
+        for a file named C(<destination>.sha256) - if that file
+        does not exist, or the hash in that file differs from
+        the local hash, the file is uploaded and the hash is
+        placed into C(<destination>.sha256) to debounce the next
+        upload.  This option requires no write access locally.
+      - C(hashcache) is the same as C(hash) except that it requires
+        local write access to C(<source>.sha256) so the hash can
+        be cached locally.  If C(source) is modified then the hash
+        will be recalculated on the next upload.  This option
+        prevents the local server from re-hashing large data files.
+      - If C(none), the file will always be uploaded and no
+        C(.sha256) files will be used.  This is appropriate
+        for smaller files.
+      default: none
+      choices: [ none, hash, hashcache ]
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+- name: Upload, overwriting any existing file at the destination.
+  vmware_datastore_upload:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    datastore: '{{ datastore_name }}'
+    source: '{{ source_file_name }}'
+    destination: '{{ destination_file_name }}'
```

Comment:
You should add real values: https://github.com/ansible/ansible/pull/47271#discussion_r226405450

Reply:
All modules in VMware space uses Jinja variables like the current implementation. I would stick to this naming scheme as this will make all modules same and readable. Adding arbitary values does not make sense to me.

Reply:
@dagwieers told me to use real values. It makes sense. Instead of `'{{ source_file_name }}'` and `'{{ destination_file_name }}'`, we should use `../folder/this-file.rpm` and `this-file.rpm`. Maybe you should take the opportunity and discuss https://github.com/ansible/ansible/issues/43433 in with the core team? I agree that we should have the same scheme across the modules.

Reply:
You should at least add real values for `datastore`, `source`, and `destination`.